rasa/utils/tensorflow/model_data.py

```diff
@@ -753,14 +764,27 @@ def balanced_data(self, data: Data, batch_size: int, shuffle: bool) -> Data:
                         for i, f in enumerate(features):
                             if len(new_data[key][sub_key]) < i + 1:
                                 new_data[key][sub_key].append([])
-                            new_data[key][sub_key][i].append(
-                                f[data_idx[index] : data_idx[index] + index_batch_size]
-                            )
-
-                data_idx[index] += index_batch_size
-                if data_idx[index] >= counts_label_ids[index]:
-                    num_data_cycles[index] += 1
-                    data_idx[index] = 0
+
+                            if balance_once:
+                                over_samples_idx = \
+                                    [k % counts_label_ids[index]
+                                        for k in range(data_idx[index], data_idx[index] + index_batch_size)]
+                                new_data[key][sub_key][i].append(f[over_samples_idx])
+                            else:
+                                new_data[key][sub_key][i].append(
+                                    f[data_idx[index] : data_idx[index] + index_batch_size]
+                                )
+
+                if balance_once:
+                    samples_per_class[index] += index_batch_size
+                    data_idx[index] += index_batch_size
+                    data_idx[index] %= counts_label_ids[index]
+                    num_data_cycles[index] = samples_per_class[index] // counts_label_ids[index]
+                else:
+                    data_idx[index] += index_batch_size
+                    if data_idx[index] >= counts_label_ids[index]:
+                        num_data_cycles[index] += 1
+                        data_idx[index] = 0
```

Comment:
This is second issue - **data_idx[index]** got reset to 0, where we could have "overlapped" samples to be pulled from the beginning of the class samples list to the training set. This only true when **index_batch_size > 1**